src/io/hdf_hyperslab.h

```diff
@@ -12,202 +12,164 @@
+  /** constructor
+   * @tparam IT integer type
+   * @param a data containter
+   * @param dims_in dimension sizes of a dataset.
+   * @param selected_in dimension sizes of the selected part of the dataset
+   * @param offsets_in offsets of the selected part of the dataset
+   *
+   * value 0 in any dimension of dims_in or selected_in is allowed when reading a dataset.
+   * The actual size is derived from file dataset.
+   */
+  template<typename IT, typename = std::enable_if_t<std::is_unsigned<IT>::value>>
+  inline hyperslab_proxy(CT& a,
+                         const std::array<IT, RANK>& dims_in,
+                         const std::array<IT, RANK>& selected_in,
+                         const std::array<IT, RANK>& offsets_in)
+      : ref_(a)
   {
-    use_slab  = false;
-    slab_rank = dims_in.size();
-    for (int i = 0; i < dims_in.size(); ++i)
-      slab_dims[i] = static_cast<hsize_t>(dims_in[i]);
-    slab_dims_local = slab_dims;
-    if (element_size > 1)
+    for (int i = 0; i < slab_rank; ++i)
     {
-      slab_dims[slab_rank] = element_size;
-      slab_rank += 1;
-    }
-  }
-
-  template<typename IC1, typename IC2, typename IC3>
-  inline hyperslab_proxy(CT& a, const IC1& dims_in, const IC2& dims_loc, const IC3& offsets_in) : container_proxy<CT>(a)
-  {
-    slab_rank = dims_in.size();
-    for (int i = 0; i < dims_in.size(); ++i)
-      slab_dims[i] = static_cast<hsize_t>(dims_in[i]);
-    for (int i = 0; i < dims_loc.size(); ++i)
-      slab_dims_local[i] = static_cast<hsize_t>(dims_loc[i]);
-    for (int i = 0; i < dims_in.size(); ++i)
+      file_space.dims[i] = static_cast<hsize_t>(dims_in[i]);
+      selected_space.dims[i] = static_cast<hsize_t>(selected_in[i]);
       slab_offset[i] = static_cast<hsize_t>(offsets_in[i]);
-    if (element_size > 1)
-    {
-      slab_dims[slab_rank]       = element_size;
-      slab_dims_local[slab_rank] = element_size;
-      slab_offset[slab_rank]     = 0;
-      slab_rank += 1;
     }
-    use_slab = true;
+
+    /// element_type related dimensions always have offset 0
+    for (int i = slab_rank; i < SpaceType::rank; ++i)
+      slab_offset[i] = 0;
   }
 
-  /** return the size of the i-th dimension
+  /** return the size of the i-th dimension of global space
    * @param i dimension
    */
-  inline hsize_t size(int i) const { return (i > MAXDIM) ? 0 : slab_dims[i]; }
-
-  inline void change_shape() { this->resize(slab_dims.data(), MAXDIM + 1); }
-};
+  inline hsize_t size(int i) const { return (i > SpaceType::rank) ? 0 : file_space.dims[i]; }
 
-template<typename CT, unsigned MAXDIM>
-struct h5data_proxy<hyperslab_proxy<CT, MAXDIM>>
-{
-  hyperslab_proxy<CT, MAXDIM>& ref_;
-  h5data_proxy(hyperslab_proxy<CT, MAXDIM>& a) : ref_(a) {}
-  inline bool read(hid_t grp, const std::string& aname, hid_t xfer_plist = H5P_DEFAULT)
+  /** checks if file_space, elected_space and offset are self-consistent
+   */
+  inline void checkUserRankSizes()
```

Comment:
@lshulen runtime checks of filespace, selected space and offset are invoked in both reading and writing. You think more checks are needed?

Reply:
My point is not that more runtime checks were needed.  It was simply that enforcing that the indices be > 0 by using an unsigned type doesn't really get you out of performing runtime checks if you want to make sure things are OK.  

Then, my logic was that if runtime checks are needed, than having one compile time check (indices are greater than 0) and several runtime checks seems a bit muddy.  Why not just make them all runtime checks at that point?

Reply:
@lshulen I tend to agree with you. Since the compiler doesn't always warning signed to unsigned conversion. A conversion from -1 to unsigned introduces a crazy number. I will remove the unsigned type and add back runtime checks.

Reply:
The above hyperslab constructor throw runtime error when negative values are detected.